src/components/structures/ScrollPanel.js

```diff
@@ -131,8 +148,53 @@ module.exports = React.createClass({
         if (sn.scrollTop < sn.clientHeight) {
             // there's less than a screenful of messages left - try to get some
             // more messages.
-            this.props.onFillRequest(true);
+            this._maybeFill(true);
         }
+        if (sn.scrollTop > sn.scrollHeight - sn.clientHeight * 2) {
+            this._maybeFill(false);
+        }
+    },
+
+    // check if there is already a pending fill request. If not, set one off.
+    _maybeFill: function(backwards) {
+        var dir = backwards ? 'b' : 'f';
+        if (this._pendingFillRequests[dir]) {
+            if (DEBUG_SCROLL) {
+                console.log("ScrollPanel: Already a "+dir+" fill in progress - not starting another");
+            }
+            return;
+        }
+
+        if (DEBUG_SCROLL) {
+            console.log("ScrollPanel: starting "+dir+" fill");
+        }
```

Comment:
Nit: It would be clearer if you could split this out to a `debuglog` function or equivalent. This would stop you having to constantly check if debug mode is on. E.g.

```javascript
function debuglog(args) {
  if (!DEBUG_SCROLL) {
    return;
  }
  // pass through all args (this allows %s to not be broken)
  console.log.apply(console, arguments);
}

// usage:
debuglog("ScrollPanel: starting "+dir+" fill");
```